anchore_engine/services/policy_engine/engine/vulns/mappers.py

```diff
@@ -929,6 +930,91 @@ def _transform_cvss(self, vulnerability_id: str, cvss: Dict) -> Dict:
+        """
+        if "nvd" in grype_vulnerability_metadata.namespace.lower():
+            return []
+        else:
+            return self._cvss_from_grype_raw_result(grype_vulnerability_metadata)
+
+    def _nvd_cvss_data(
+        self,
+        grype_raw_result,
+        nvd_cvss_cache: dict,
+        related_nvd_metadata_records: List[GrypeVulnerabilityMetadata],
+    ) -> List[dict]:
+        """
+        Builds nvd data for provided grype vulnerability metadata record
+        Uses a cache dict param that maps a vuln id to nvd cvss data to avoid needing to search existing query results
+        or creating a new query to find nvd metadata record for each vulnerability
+        """
+        nvd_data = []
+        grype_vulnerability_metadata = grype_raw_result.GrypeVulnerabilityMetadata
+        grype_vulnerability = grype_raw_result.GrypeVulnerability
+
+        # If current record is nvd, get cvss data for record
+        if "nvd" in grype_vulnerability_metadata.namespace.lower():
+            # Check in cache before building out from grype record
+            if nvd_cvss_cache.get(grype_vulnerability_metadata.id):
+                nvd_data += nvd_cvss_cache.get(grype_vulnerability_metadata.id)
+            else:
+                raw_record_nvd_data = self._cvss_from_grype_raw_result(
+                    grype_vulnerability_metadata
+                )
+                nvd_cvss_cache[grype_vulnerability_metadata.id] = raw_record_nvd_data
+                nvd_data += raw_record_nvd_data
+
+        # get cvss data for related vulnerabilities to populate
+        if grype_vulnerability.deserialized_related_vulnerabilities:
+            for (
+                related_vuln
+            ) in grype_vulnerability.deserialized_related_vulnerabilities:
+                if "nvd" not in related_vuln["Namespace"].lower():
+                    continue
+
+                # Check cache, else find nvd record and build cvss data
+                if nvd_cvss_cache.get(related_vuln["ID"]):
+                    nvd_data += nvd_cvss_cache.get(related_vuln["ID"])
+                else:
+                    nvd_metadata_record = next(
+                        (
+                            nvd_metadata_record
+                            for nvd_metadata_record in related_nvd_metadata_records
+                            if nvd_metadata_record.id == related_vuln["ID"]
+                        ),
+                        None,
+                    )
+
+                    if nvd_metadata_record:
+                        related_vuln_nvd_data = self._cvss_from_grype_raw_result(
+                            nvd_metadata_record
+                        )
+                        nvd_cvss_cache[nvd_metadata_record.id] = related_vuln_nvd_data
+                        nvd_data += related_vuln_nvd_data
+                    else:
+                        log.debug("Missing related record for %s", related_vuln["ID"])
+
```

Comment:
This seems a little complicated. Suggestions for simplification:
- build a map of nvd-id->cvss score map prior to this call and use that for look ups. It might be more performant/efficient than the cache built here
- the logic is a bit convoluted for nvd and non-nvd vulnerabilities, that's another thing that could be handled in a higher order layer